dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java

```diff
@@ -215,39 +215,33 @@ public void endInitialization(WebSphereCDIDeployment webSphereCDIDeployment) thr
         if (eventManager != null) {
-            WebSphereCDIDeployment deployment = getDeployment(application);
-            if (deployment != null) {
-                BeanDeploymentModules modules = deployment.getServices().get(BeanDeploymentModules.class);
-                if (modules != null) {
-                    for (BeanDeploymentModule module : modules) {
-                        if (!module.isWebModule()) {
-                            String id = module.getId();
-                            WebSphereBeanDeploymentArchive bda = deployment.getBeanDeploymentArchive(id);
-                            if (bda != null) {
-                                try (ContextBeginnerEnder contextBeginnerEnder = cdiRuntime.createContextBeginnerEnder().extractComponentMetaData(bda.getArchive())
-                                                                                           .extractTCCL(application).beginContext()) {
-                                    eventManager.fireStartupEvent(module);
-                                }
-                            } else {
-                                throw new IllegalStateException(Tr.formatMessage(tc, "no.bda.for.module.CWOWB1019E", module, module.getId()));
-                            }
-                        }
-                    }
-                }
-            }
+            fireEventWithContext(eventManager::fireStartupEvent, application, eventManager);
         }
     }
 
     public void applicationStopping(Application application) throws CDIException {
         CDIContainerEventManager eventManager = cdiRuntime.getCDIContainerEventManager();
         if (eventManager != null) {
-            WebSphereCDIDeployment deployment = getDeployment(application);
-            if (deployment != null) {
-                BeanDeploymentModules modules = deployment.getServices().get(BeanDeploymentModules.class);
-                if (modules != null) {
-                    for (BeanDeploymentModule module : modules) {
-                        if (!module.isWebModule()) {
-                            eventManager.fireShutdownEvent(module);
+            fireEventWithContext(eventManager::fireShutdownEvent, application, eventManager);
+        }
+    }
+
+    private void fireEventWithContext(Consumer<BeanDeploymentModule> event, Application application, CDIContainerEventManager eventManager) throws CDIException {
+        WebSphereCDIDeployment deployment = getDeployment(application);
+        if (deployment != null) {
+            BeanDeploymentModules modules = deployment.getServices().get(BeanDeploymentModules.class);
+            if (modules != null) {
+                for (BeanDeploymentModule module : modules) {
+                    if (!module.isWebModule()) {
+                        String id = module.getId();
+                        WebSphereBeanDeploymentArchive bda = deployment.getBeanDeploymentArchive(id);
```

Comment:
I don't like this way of getting a BDA for a module, it seems very breakable and doesn't match the way `BeanDeploymentModules` navigates from a `BeanManagerImpl` to a `BeanDeploymentModule`.

Probably warrants a tech debt item.

Reply:
I opened https://github.com/OpenLiberty/open-liberty/issues/29279